lib/csv_importer/csv_reader.rb

```diff
@@ -53,7 +53,18 @@ def sanitize_content(csv_content)
     SEPARATORS = [",", ";", "\t"]
 
     def detect_separator(csv_content)
-      SEPARATORS.sort_by { |separator| csv_content.count(separator) }.last
+      SEPARATORS.min_by do |separator|
+        csv_content.count(separator)
+
+        all_lines = csv_content.lines
+        base_number = all_lines.first.count(separator)
+
+        if base_number.zero?
+          Float::MAX
+        else
+          all_lines.map{|line| line.count(separator) - base_number }.map(&:abs).inject(0) { |sum, i| sum + i }
```

Comment:
You compare each line against the header line. That works as long as the header doesn't contain too many of the wrong separator. :)

I'm wondering if another probability / stats method could give us better results. Variance for example?

Or maybe electing the separator with the greatest number of lines with the same number of separators? Would that break down for CSVs with only 1 entry?

Anyway, I'm happy to go with this as it's already better than what we have. :)

Reply:
> That works as long as the header doesn't contain too many of the wrong separator. :)

Actually, look closer. It's a scoring mechanism - through a simple one.

We assume that a 'correct' CSV file has the same count of separators in each line. In this line we calculate 'inconsistencies' and put a point for each inconsistency in every line. And the 'correct/valid' CSV will have a score of 0, while on non-specially-prepared-data will have more then one. i.e. if we have 3 commas in header, 2 in one line and 5 in another - score for comma would be 3 ( `abs(3-2) + abs(3-5) = 1 + 2 = 3` ).

And we take the one which has the least score. It's definitely better then existing one - surely you can still come up with counterexample, but that would be covered by model-level or header validations in real world.